encoder.go

```diff
@@ -0,0 +1,162 @@
+package shortuuid
+
+import (
+	"encoding/binary"
+	"fmt"
+	"github.com/google/uuid"
+	"math"
+	"math/bits"
+	"strings"
+)
+
+type encoder struct {
+	// alphabet is the character set to construct the UUID from.
+	alphabet alphabet
+}
+
+const (
+	defaultBase    = 57
+	defaultEncLen  = 22
+	defaultNDigits = 10
+	defaultDivisor = 362033331456891249 // 57^10
+)
+
+func maxPow(b uint64) (d uint64, n int) {
+	d, n = b, 1
+	for m := math.MaxUint64 / b; d <= m; {
+		d *= b
+		n++
+	}
+	return
+}
+
+// Encode encodes uuid.UUID into a string using the most significant bits (MSB)
+// first according to the alphabet.
+func (e encoder) Encode(u uuid.UUID) string {
+	if e.alphabet.singleBytes {
+		return e.encodeSingleBytes(u)
+	}
+	return e.encode(u)
+}
+
+func (e encoder) encodeSingleBytes(u uuid.UUID) string {
+	num := uint128{
+		binary.BigEndian.Uint64(u[8:]),
+		binary.BigEndian.Uint64(u[:8]),
+	}
+	var r uint64
+	var i int
+	var buf []byte
+	if e.alphabet.len == defaultBase { // compiler optimizations using constants for default base
+		buf = make([]byte, defaultEncLen)
+		for i = defaultEncLen - 1; num.Hi > 0 || num.Lo > 0; {
+			num, r = num.quoRem64(defaultDivisor)
+			for j := 0; j < defaultNDigits && i >= 0; j++ {
+				buf[i] = byte(e.alphabet.chars[r%defaultBase])
+				r /= defaultBase
+				i--
+			}
+		}
+	} else {
+		buf = make([]byte, e.alphabet.encLen)
+		l := uint64(e.alphabet.len)
+		d, n := maxPow(l)
+		for i = int(e.alphabet.encLen - 1); num.Hi > 0 || num.Lo > 0; {
+			num, r = num.quoRem64(d)
+			for j := 0; j < n && i >= 0; j++ {
+				buf[i] = byte(e.alphabet.chars[r%l])
+				r /= l
+				i--
+			}
+		}
+	}
+	for ; i >= 0; i-- {
+		buf[i] = byte(e.alphabet.chars[0])
+	}
+	return string(buf[:])
+}
+
+func (e encoder) encode(u uuid.UUID) string {
+	num := uint128{
+		binary.BigEndian.Uint64(u[8:]),
+		binary.BigEndian.Uint64(u[:8]),
+	}
+	var r uint64
+	var outIndexes []uint64
+	if e.alphabet.len == defaultBase { // compiler optimizations using constants for default base
```

Comment:
How big is the difference? I'm just thinking of the added complexity is really worth it?

Reply:
From my benchmarks, it's pretty significant.
~ +127% on `encodeSingleBytes`
~ +74% on `encode`
~ +11% on `Decode`

Here is full benchmark:
```
goos: linux
goarch: amd64
cpu: 12th Gen Intel(R) Core(TM) i5-12400F
                         │     new     │              not-optimized              │
                         │   sec/op    │    sec/op     vs base                   │
UUID-12                    332.3n ± 0%    385.9n ± 1%   +16.15% (p=0.000 n=10)
Encoding-12                44.73n ± 0%   101.60n ± 0%  +127.14% (p=0.000 n=10)
EncodingB57_MB-12          118.8n ± 2%    206.6n ± 3%   +73.91% (p=0.000 n=10)
EncodingB16-12             134.8n ± 2%    134.4n ± 0%    -0.30% (p=0.024 n=10)
EncodingB16_MB-12          329.2n ± 0%    330.8n ± 1%    +0.47% (p=0.022 n=10)
Decoding-12                176.7n ± 0%    196.3n ± 0%   +11.03% (p=0.000 n=10)
DecodingB16-12             215.4n ± 1%    242.8n ± 1%   +12.72% (p=0.000 n=10)
DecodingB16_MB-12          256.8n ± 2%    284.3n ± 1%   +10.75% (p=0.000 n=10)
NewWithAlphabet-12         624.6n ± 1%    666.0n ± 0%    +6.62% (p=0.000 n=10)
NewWithAlphabetB16-12      495.0n ± 1%    505.9n ± 0%    +2.21% (p=0.000 n=10)
NewWithAlphabetB16_MB-12   809.7n ± 1%    812.1n ± 0%         ~ (p=0.109 n=10)
geomean                    245.9n         292.4n        +19.29%                

                         │     new      │              not-optimized               │
                         │     B/op     │    B/op      vs base                     │
UUID-12                    40.00 ± 0%      64.00 ± 0%   +60.00% (p=0.000 n=10)
Encoding-12                24.00 ± 0%      48.00 ± 0%  +100.00% (p=0.000 n=10)
EncodingB57_MB-12          72.00 ± 0%     248.00 ± 0%  +244.44% (p=0.000 n=10)
EncodingB16-12             64.00 ± 0%      64.00 ± 0%         ~ (p=1.000 n=10)
EncodingB16_MB-12          480.0 ± 0%      480.0 ± 0%         ~ (p=1.000 n=10)
Decoding-12                0.000 ± 0%      0.000 ± 0%         ~ (p=1.000 n=10)
DecodingB16-12             0.000 ± 0%      0.000 ± 0%         ~ (p=1.000 n=10)
DecodingB16_MB-12          0.000 ± 0%      0.000 ± 0%         ~ (p=1.000 n=10)
NewWithAlphabet-12         280.0 ± 0%      304.0 ± 0%    +8.57% (p=0.000 n=10)
NewWithAlphabetB16-12      144.0 ± 0%      144.0 ± 0%         ~ (p=1.000 n=10)
NewWithAlphabetB16_MB-12   560.0 ± 0%      560.0 ± 0%         ~ (p=1.000 n=10)
geomean                                                 +25.31%                

                         │     new      │              not-optimized              │
                         │  allocs/op   │ allocs/op   vs base                     │
UUID-12                    2.000 ± 0%     3.000 ± 0%   +50.00% (p=0.000 n=10)
Encoding-12                1.000 ± 0%     2.000 ± 0%  +100.00% (p=0.000 n=10)
EncodingB57_MB-12          2.000 ± 0%     3.000 ± 0%   +50.00% (p=0.000 n=10)
EncodingB16-12             2.000 ± 0%     2.000 ± 0%         ~ (p=1.000 n=10)
EncodingB16_MB-12          4.000 ± 0%     4.000 ± 0%         ~ (p=1.000 n=10)
Decoding-12                0.000 ± 0%     0.000 ± 0%         ~ (p=1.000 n=10)
DecodingB16-12             0.000 ± 0%     0.000 ± 0%         ~ (p=1.000 n=10)
DecodingB16_MB-12          0.000 ± 0%     0.000 ± 0%         ~ (p=1.000 n=10)
NewWithAlphabet-12         3.000 ± 0%     4.000 ± 0%   +33.33% (p=0.000 n=10)
NewWithAlphabetB16-12      4.000 ± 0%     4.000 ± 0%         ~ (p=1.000 n=10)
NewWithAlphabetB16_MB-12   6.000 ± 0%     6.000 ± 0%         ~ (p=1.000 n=10)
geomean                                                +17.69%                
```